Sources/TuistCore/Graph/GraphLoaderLinter.swift

```diff
@@ -0,0 +1,128 @@
+import Foundation
+import TSCBasic
+import TuistGraph
+
+// MARK: - GraphLoaderLinting
+
+public protocol GraphLoaderLinting {
+    func lintWorkspace(workspace: Workspace, projects: [Project]) throws
+    func lintProject(at path: AbsolutePath, projects: [Project]) throws
+}
+
+// MARK: - GraphLoaderLinter
+
+public class GraphLoaderLinter {
+    public init() {}
+
+    public func lintWorkspace(workspace: Workspace, projects: [Project]) throws {
+        let cycleDetector = GraphCircularDetector()
+        let cache = try GraphLoader.Cache(projects: projects)
+        try workspace.projects.forEach {
+            try lintProject(
+                path: $0,
+                cache: cache,
+                cycleDetector: cycleDetector
+            )
+        }
+    }
+
+    public func lintProject(at path: AbsolutePath, projects: [Project]) throws {
+        let cache = try GraphLoader.Cache(projects: projects)
```

Comment:
I find creating a cache here very misleading. In the context of loading the graph it makes sense because you want to avoid unnecessary loads, but once the graph instance is created, we should be using that one instead. What's the rationale behind using a cache model that's scoped to the loading phase?

Reply:
> but once the graph instance is created, we should be using that one instead.
Not sure what you are referring here. I am using the cache to avoid passing from a node which we already visit, in fact I believe it will also avoid getting into an infinite loop when having a circular dependency, isn't?